hypothesis/client
# Add initial implementation of Import tab in share panel

Add an initial implementation of the UI for the Import tab in the share panel, along with a supporting service to perform the import, and store fields to track progress.

Part of https://github.com/hypothesis/client/issues/5695, https://github.com/hypothesis/client/issues/5693, https://github.com/hypothesis/client/issues/5697.

## Summary of UI changes

When the user clicks the "Share" icon in the top bar and selects the Import tab, and the user is logged in, they will see:

<img width="425" alt="Import tab initial state" src="https://github.com/hypothesis/client/assets/2458/054e0762-5410-4c61-b5df-77a1a3a352b2">

If the user is logged out, they will just see a notice that they need to login first:

<img width="428" alt="Import tab login notice" src="https://github.com/hypothesis/client/assets/2458/2ef7ed76-70d5-4e5b-8cc1-cf7637169f36">

After selecting a file, they will be shown an alphabetized list of users and annotation counts. If the current user ID matches a user in the list, that user will be preselected:

<img width="425" alt="Import tab user list" src="https://github.com/hypothesis/client/assets/2458/a47b2141-e8ef-4802-973e-1d91f7d01b25">

Clicking the "Import" button starts the import and shows a spinner. After a successful import, a toast is shown:

<img width="420" alt="Import ok" src="https://github.com/hypothesis/client/assets/2458/861c0798-71ac-41ed-9b4b-2a1d9f581c6a">

Annotations which are duplicates of existing annotations are skipped. This will be indicated in the toast after the import:

<img width="421" alt="Import skipped" src="https://github.com/hypothesis/client/assets/2458/efd2a33c-87f7-40ee-8073-3aa5ba9b07da">

This toast will be orange if all imports were skipped, or green if only some were:

<img width="432" alt="Some dupes" src="https://github.com/hypothesis/client/assets/2458/52e58713-fdfd-4978-a352-886d15e9a3eb">

If failures occur, the user will see a red toast if all imports (ignoring duplicates) failed, or an orange toast if only some did:

<img width="437" alt="All failed" src="https://github.com/hypothesis/client/assets/2458/f5291f39-d944-4f66-8e43-d07b8999d51d">

## Summary of code changes

- Copy the latest version of the `wait*` testing utilities from the Via repo. We should make these into a new shared package soon.
- Add an `importsPending` field to the activity store module, along with `beginImport` and `completeImport` actions to signal when imports begin and end
- Add `ImportAnnotationsService` which performs the import
- Test for `import_annotations` feature flag and show Import tab in `ShareDialog` when it is enabled
- Add `ImportAnnotations` component which implements the UI for the "Import" tab

## Caveats

Caveats of this initial implementation:

- The styling of the Import tab is not finished. In particular the file input doesn't have the custom styling of the mocks.
- Validation of the JSON data is very minimal ("is it valid JSON, does it have an `annotations` property that is an array?"). See also ([this slack thread](https://hypothes-is.slack.com/archives/C1M8NH76X/p1691595960140369))
- De-duplication is currently based on the content of annotations (text + tags + quote), but this is designed to be easy to change in future. One reason in this initial PR is that it doesn't require storing any additional data with imported annotations.
- Annotations are imported serially. Once the initial version lands, we can look at implementing batching
- The URLs of imported annotations are not checked against the current document or modified to match the current document. If the user imports annotations for a different URL, they will appear in the sidebar after the initial import, but will disappear after a page reload, since the annotations belong to a different URL.
- All imported annotations are currently set to shared

## Testing

Make sure the `export_annotations` and `import_annotations` feature flags are enabled.

1. Go to any page that has annotations, and use the Export tab to export annotations to a file
2. Switch to a different group, and select the file in the Import tool. A correct count of users and annotations should appear. If you authored any of the annotations, your name should be pre-selected in the user list, and you can change this, otherwise you will be required to select a user.
3. Click Import. The annotations should be imported to the current group. The username, date and other properties of annotations which are set automatically should be the same as if you created the annotation locally.
4. Try to repeat the import again. No new annotations should be imported and a notice should be shown indicating that all the selected annotations are duplicates of those already in the group.
5. Delete one or two annotations and try to import again. The deleted annotations will be replaced, but the existing ones will not be duplicated.
